### PR TITLE
Enforce consistent naming for agent threads

### PR DESCRIPTION
**Pull request by kananindzya**
### Checklist

closes elastic/apm-agent-java#960 
- [x] Implement code
- [x] Add tests
- [ ] Update documentation

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/SoftlyReferencingTypePoolCache.java [hunk 1]
```diff
@@ -61,7 +62,7 @@
     public SoftlyReferencingTypePoolCache(final TypePool.Default.ReaderMode readerMode,
                                           final int clearIfNotAccessedSinceMinutes, ElementMatcher.Junction<ClassLoader> ignoredClassLoaders) {
         super(readerMode);
-        ExecutorUtils.createSingleThreadSchedulingDeamonPool("type-cache-pool-cleaner", 1)
+        ExecutorUtils.createSingleThreadSchedulingDeamonPool(ThreadUtils.addElasticApmThreadPrefix("type-cache-pool-cleaner"), 1)
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/SoftlyReferencingTypePoolCache.java]: fixed

**Comment by kananindzya** [comment]
Thread prefixes:
- metrics-reporter 
- type-cache-pool-cleaner 
- remote-config-poller
- server-healthcheck 
- server-reporter
- init-instrumentation-shutdown-hook
- request-timeout-timer

Logs example:
```
2020-01-15 17:43:27.628 [elastic-apm-server-healthcheck] DEBUG co.elastic.apm.agent.report.ApmServerHealthChecker - Starting healthcheck to http://localhost:8200/
2020-01-15 17:43:27.637 [elastic-apm-server-healthcheck] INFO co.elastic.apm.agent.report.ApmServerHealthChecker - Elastic APM server is available: {  "build_date": "2019-12-16T20:57:59Z",  "build_sha": "348d8d83c3c823b64fc0692be607b1a5a8fac775",  "version": "7.5.1"}
2020-01-15 17:43:27.638 [main] DEBUG co.elastic.apm.agent.impl.payload.SystemInfo - Could not parse container ID from '/proc/self/cgroup' line: 12:memory:/user.slice/user-1002.slice/user@1002.service
2020-01-15 17:43:27.639 [main] DEBUG co.elastic.apm.agent.impl.payload.SystemInfo - Could not parse container ID from '/proc/self/cgroup' line: 11:rdma:/
2020-01-15 17:43:27.639 [elastic-apm-server-healthcheck] DEBUG co.elastic.apm.agent.report.ApmServerHealthChecker - APM server http://localhost:8200/ version is: 7.5.1


2020-01-15 17:43:27.668 [elastic-apm-remote-config-poller] DEBUG co.elastic.apm.agent.configuration.ApmServerConfigurationSource - Reloading configuration from APM Server http://localhost:8200/config/v1/agents
2020-01-15 17:43:27.668 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - application_packages: 'com.kananinja.apm.spring' (source: elasticapm.properties)
2020-01-15 17:43:27.668 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - log_level: 'DEBUG' (source: elasticapm.properties)
2020-01-15 17:43:27.670 [elastic-apm-remote-config-poller] DEBUG co.elastic.apm.agent.configuration.ApmServerConfigurationSource - Central configuration is disabled. Set kibana.enabled: true in your APM Server configuration.

2020-01-15 17:43:57.669 [elastic-apm-server-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Scheduling request timeout in 10s
2020-01-15 17:44:07.670 [elastic-apm-request-timeout-timer] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Request flush because the request timeout occurred
2020-01-15 17:44:07.673 [elastic-apm-server-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving FLUSH event (sequence 1)
2020-01-15 17:44:07.674 [elastic-apm-server-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Flushing 1391 uncompressed 606 compressed bytes
2020-01-15 17:44:27.669 [elastic-apm-server-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving METRICS event (sequence 2)
2020-01-15 17:44:27.670 [elastic-apm-server-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Starting new request to http://localhost:8200/intake/v2/events
2020-01-15 17:44:27.671 [elastic-apm-server-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Scheduling request timeout in 10s
```
screenshot from jconsole:
threads-elastic-apm

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/SoftlyReferencingTypePoolCache.java [hunk 2]
```diff
@@ -25,6 +25,7 @@
 package co.elastic.apm.agent.bci.bytebuddy;
 
 import co.elastic.apm.agent.util.ExecutorUtils;
+import co.elastic.apm.agent.util.ThreadUtils;
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/SoftlyReferencingTypePoolCache.java]: remove unused import

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java [hunk 3]
```diff
@@ -30,6 +30,7 @@
 import co.elastic.apm.agent.report.ApmServerClient;
 import co.elastic.apm.agent.report.serialize.PayloadSerializer;
 import co.elastic.apm.agent.util.ExecutorUtils;
+import co.elastic.apm.agent.util.ThreadUtils;
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java]: remove unused import

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerHealthChecker.java [hunk 4]
```diff
@@ -25,6 +25,7 @@
 package co.elastic.apm.agent.report;
 
 import co.elastic.apm.agent.util.ExecutorUtils;
+import co.elastic.apm.agent.util.ThreadUtils;
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerHealthChecker.java]: remove unused import

**Comment by felixbarny** [comment]
Also, there's a `weak-ref-cleaner-0` thread which seems to be created by the agent.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/SoftlyReferencingTypePoolCache.java (hunk 2 above)
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/SoftlyReferencingTypePoolCache.java]: removed

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java (hunk 3 above)
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java]: removed

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerHealthChecker.java (hunk 4 above)
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerHealthChecker.java]: removed

**Comment by kananindzya** [comment]
> Also, there's a `weak-ref-cleaner-0` thread which seems to be created by the agent.

I found that it created by WeakConcurrentSet. Added method to DataStructures#createWeakConcurrentSetWithCleanerThread

**Comment by kananindzya** [comment]
> In the Jconsole screenshot I saw another thread which is not renamed yet: `configuration-reloader`. That's a bit more tricky as the thread is created in `org.stagemonitor.configuration.ConfigurationRegistry#initThreadPool`. To work around that, instead of calling `org.stagemonitor.configuration.ConfigurationRegistry#scheduleReloadAtRate`, we could create a properly named scheduled executor service ourselves within `co.elastic.apm.agent.impl.ElasticApmTracerBuilder#build` and register a `LifecycleListener` which shuts down the thread pool.

@felixbarny I couldn't find way to pass ScheduledExecutorService to ConfigurationRegistry

**Comment by felixbarny** [comment]
This should work: https://github.com/kananindzya/apm-agent-java/pull/3

**Comment by kananindzya** [comment]
Updated screenshot:
elastic-apm-agent-threads

**Comment by kananindzya** [comment]
> This should work: [kananindzya#3](https://github.com/kananindzya/apm-agent-java/pull/3)

Awesome! Thanks

**Pull request merged**